Support multiple targets/precursors in RGDR

BSchilperoort opened the pull request:
This PR adds the possibility to specify multiple target periods for RGDR, e.g.:
```py
RGDR(
    target_intervals=[1, 2, 3],
    lag=3,
    ...
)
```
To be able to apply these changes the functionality that allows users to apply RGDR to a whole set of precursor intervals, and then analyze them separately, had to be removed. Specifying the interval indices is now required when initiating `RGDR`, and when calling `.transform` the to-be-transformed data needs to have the right precursor intervals.

- Implements #151 
- Closes #56
- Small improvements in RGDR class (added `@prop`s, `__repr__`, ...)
- Some RGDR preview plot improvements
- Corrects docstring in `s2spy.rgdr.label_alignment.rename_labels()` (based on previous comment of Peter).
- Fixes a typing issue in `s2spy.time.resample` (using `@typing.overload`)

semvijverberg commented on s2spy/rgdr/rgdr.py:
```diff
@@ -355,91 +355,132 @@ def __init__(
+        self._area = None
+
+    @property
+    def target_intervals(self) -> List[int]:
```
For my own learning experience, can I ask what the added value of this function is? self._target_intervals is already a list (defined at __init__), so why create this function and not simply name:

```
        self.target_intervals = (
             [target_intervals]
             if isinstance(target_intervals, int)
             else target_intervals
         )
```

semvijverberg replied:
Maybe it's for consistency reasons? Later on I see the added value of calling the method cluster map for example.

BSchilperoort replied:
The `@property` decorator is so we can give users information without allowing them to change anything. It is not a function in the way that users call it like a class property, e.g.:
```py
rgdr.target_intervals
```
And not:
```py
rgdr.target_intervals()  # raises an error
```

We do not want to allow users to change properties of the RGDR class after it is initiated. The underscore at the beginning of the variable name `self._target_intervals` denotes that users _should not_ touch this, and if they do so anyway that they're responsible for any weird behavior.

If we do not implement `@property`, a user would be able to do:
```py
rgdr.target_intervals = 5
```
And the lag & precursor intervals would not be affected.

If it is a `@property`, and no `@target_intervals.setter` has been implemented, Python will raise an error when a user tries to modify the value.

This is also the reason why I made the stored maps properties.

semvijverberg reviewed (approved):
Amazing work @BSchilperoort! Thanks! 

Super nicely documented, tested, and coded.

Peter9192 commented on notebooks/tutorial_RGDR.ipynb:
```diff
@@ -78,72 +84,92 @@
+    "RGDR requires `DataArray`s as input, as otherwise it is ambiguous which variable should be used.\n",
+    "\n",
+    "When initializing RGDR, the target interval(s) and lag have to be specified as well.\n",
+    "The precursor interval(s) will be determined from the target intervals and lag.\n",
```
Add a (more explicit) note here that the maximum lag is constrained by the range of the available precursor intervals?

BSchilperoort replied:
Yes, good point. It is now explicitly mentioned.

Peter9192 reviewed (approved):
I briefly looked over the code and examples, looks very good! The modified functionality makes sense to me.

I was wondering whether the example illustration for multiple targets also has a source file (for easy adaptation later on). Also, in the example notebook, I think it would be helpful to have a similar explanatory figure for the single target interval.

BSchilperoort commented on s2spy/rgdr/label_alignment.py:
```diff
@@ -358,15 +357,14 @@ def _rename_datasets(
         data = copy(clustered_data[split])
-        labels = data["cluster_labels"].values
-        i_interval = str(_rgdr.cluster_map["i_interval"].values)  # type: ignore
+        labels = data["cluster_labels"].values.astype("U32")
```
In xarray you cannot have arbitrarily long strings as coordinates, so I chose for a 32 character Unicode string, as that would fit well enough if users would want to rename clusters to something more descriptive like "north-west pacific ocean".

Otherwise the name would be trimmed if users try to change values in the array. E.g.:
![image](https://user-images.githubusercontent.com/12114825/211825640-1259ebc6-c072-4a42-986a-1b10f24801fe.png)

BSchilperoort commented:
> I briefly looked over the code and examples, looks very good! The modified functionality makes sense to me.
> 
> I was wondering whether the example illustration for multiple targets also has a source file (for easy adaptation later on). Also, in the example notebook, I think it would be helpful to have a similar explanatory figure for the single target interval.

Thanks for the review, Peter. Good point about adding the source file (and illustration for the single-target case). I have converted the ppt file to svg files, and added the single-target illustration.
